File: app/src/main/java/org/mifos/mobilebanking/presenters/HomePresenter.java
package org.mifos.mobilebanking.presenters;

import android.content.Context;
import android.graphics.Bitmap;
import android.util.Base64;
import android.util.Log;

import org.mifos.mobilebanking.R;
import org.mifos.mobilebanking.api.DataManager;
import org.mifos.mobilebanking.api.local.PreferencesHelper;
import org.mifos.mobilebanking.injection.ApplicationContext;
import org.mifos.mobilebanking.models.client.Client;
import org.mifos.mobilebanking.presenters.base.BasePresenter;
import org.mifos.mobilebanking.ui.views.HomeView;
import org.mifos.mobilebanking.utils.ImageUtil;

import java.io.IOException;

import javax.inject.Inject;

import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.observers.DisposableObserver;
import io.reactivex.schedulers.Schedulers;
import okhttp3.ResponseBody;

/**
 * Created by dilpreet on 19/6/17.
 */

public class HomePresenter extends BasePresenter<HomeView> {

    private DataManager dataManager;
    private CompositeDisposable compositeDisposable;
    @Inject
    PreferencesHelper preferencesHelper;

    /**
     * Initialises the LoginPresenter by automatically injecting an instance of
     * {@link DataManager} and {@link Context}.
     *
     * @param dataManager DataManager class that provides access to the data
     *                    via the API.
     * @param context     Context of the view attached to the presenter. In this case
     *                    it is that of an {@link android.support.v7.app.AppCompatActivity}
     */
    @Inject
    public HomePresenter(DataManager dataManager, @ApplicationContext Context context) {
        super(context);
        this.dataManager = dataManager;
        compositeDisposable = new CompositeDisposable();
    }

    @Override
    public void attachView(HomeView mvpView) {
        super.attachView(mvpView);
    }

    @Override
    public void detachView() {
        super.detachView();
        compositeDisposable.clear();
    }

    /**
     * Fetches Details about Client from the server as {@link Client} and notifies the view to
     * display the details. And in case of any error during fetching the required details it
     * notifies the view.
     */
    public void getUserDetails() {
        checkViewAttached();
        compositeDisposable.add(dataManager.getCurrentClient()
                .observeOn(AndroidSchedulers.mainThread())
                .subscribeOn(Schedulers.io())
                .subscribeWith(new DisposableObserver<Client>() {
                    @Override
                    public void onComplete() {

                    }

                    @Override
                    public void onError(Throwable e) {
                        getMvpView().showError(context.getString(R.string.error_fetching_client));
                        getMvpView().hideProgress();
                    }

                    @Override
                    public void onNext(Client client) {
                        getMvpView().hideProgress();
                        if (client != null) {
                            preferencesHelper.setOfficeName(client.getOfficeName());
                            preferencesHelper.setClientName(client.getDisplayName());
                            getMvpView().showUserDetails(preferencesHelper.getClientName());
                        } else {
                            getMvpView().showError(context
                                    .getString(R.string.error_client_not_found));
                        }
                    }
                })
        );
    }

    /**
     * Fetches Client image from the server in {@link Base64} format which is then decoded into a
     * {@link Bitmap} after which the view notified to display it.
     */
    public void getUserImage() {
        checkViewAttached();
        compositeDisposable.add(dataManager.getClientImage()
                .observeOn(Schedulers.newThread())
                .subscribeOn(Schedulers.io())
                .subscribeWith(new DisposableObserver<ResponseBody>() {
                    @Override
                    public void onComplete() {

                    }

                    @Override
                    public void onError(Throwable e) {
                        getMvpView().showUserImageNotFound();
                    }

                    @Override
                    public void onNext(ResponseBody response) {
                        try {

                            final String encodedString = response.string();

                            final String pureBase64Encoded =
                                    encodedString.substring(encodedString.indexOf(',') + 1);

                            final byte[] decodedBytes =
                                    Base64.decode(pureBase64Encoded, Base64.DEFAULT);

                            Bitmap decodedBitmap = ImageUtil.getInstance().
                                    compressImage(decodedBytes, 256, 256);
                            getMvpView().showUserImage(decodedBitmap);
                        } catch (IOException e) {
                            Log.d("userimage", e.toString());
                        }
                    }
                })
        );
    }

    public void getUnreadNotificationsCount() {
        compositeDisposable.add(dataManager.getUnreadNotificationsCount()
                .observeOn(AndroidSchedulers.mainThread())
                .subscribeOn(Schedulers.computation())
                .subscribeWith(new DisposableObserver<Integer>() {
                    @Override
                    public void onComplete() {

                    }

                    @Override
                    public void onError(Throwable e) {

                    }

                    @Override
                    public void onNext(Integer integer) {
                        getMvpView().showNotificationCount(integer);
                    }
                }));
    }
}

File: app/src/main/java/org/mifos/mobile/presenters/HomePresenter.kt
package org.mifos.mobile.presenters

import android.content.Context
import android.graphics.Bitmap
import android.util.Base64
import android.util.Log

import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.observers.DisposableObserver
import io.reactivex.schedulers.Schedulers

import okhttp3.ResponseBody

import org.mifos.mobile.R
import org.mifos.mobile.api.DataManager
import org.mifos.mobile.api.local.PreferencesHelper
import org.mifos.mobile.injection.ApplicationContext
import org.mifos.mobile.models.client.Client
import org.mifos.mobile.presenters.base.BasePresenter
import org.mifos.mobile.ui.views.HomeView
import org.mifos.mobile.utils.ImageUtil

import java.io.IOException
import javax.inject.Inject

/**
 * Created by dilpreet on 19/6/17.
 */
class HomePresenter @Inject constructor(private val dataManager: DataManager, @ApplicationContext context: Context?) : BasePresenter<HomeView?>(context) {
    private val compositeDisposable: CompositeDisposable? = CompositeDisposable()

    @JvmField
    @set:Inject
    var preferencesHelper: PreferencesHelper? = null
    override fun attachView(mvpView: HomeView?) {
        super.attachView(mvpView)
    }

    override fun detachView() {
        super.detachView()
        compositeDisposable?.clear()
    }

    /**
     * Fetches Details about Client from the server as [Client] and notifies the view to
     * display the details. And in case of any error during fetching the required details it
     * notifies the view.
     */
    val userDetails: Unit
        get() {
            checkViewAttached()
            dataManager.currentClient
                    ?.observeOn(AndroidSchedulers.mainThread())
                    ?.subscribeOn(Schedulers.io())
                    ?.subscribeWith(object : DisposableObserver<Client?>() {
                        override fun onComplete() {}
                        override fun onError(e: Throwable) {
                            mvpView?.showError(context?.getString(R.string.error_fetching_client))
                            mvpView?.hideProgress()
                        }

                        override fun onNext(client: Client) {
                            mvpView?.hideProgress()
                            preferencesHelper?.officeName = client.officeName
                            preferencesHelper?.clientName = client.displayName
                            mvpView?.showUserDetails(preferencesHelper?.clientName)
                        }
                    })?.let {
                        compositeDisposable?.add(it
                        )
                    }
        }

    /**
     * Fetches Client image from the server in [Base64] format which is then decoded into a
     * [Bitmap] after which the view notified to display it.
     */
    val userImage: Unit
        get() {
            checkViewAttached()
            dataManager.clientImage
                    ?.observeOn(Schedulers.newThread())
                    ?.subscribeOn(Schedulers.io())
                    ?.subscribeWith(object : DisposableObserver<ResponseBody?>() {
                        override fun onComplete() {}
                        override fun onError(e: Throwable) {
                            mvpView?.showUserImageNotFound()
                        }

                        override fun onNext(response: ResponseBody) {
                            try {
                                val encodedString = response.string()
                                val pureBase64Encoded = encodedString.substring(encodedString.indexOf(',') + 1)
                                val decodedBytes = Base64.decode(pureBase64Encoded, Base64.DEFAULT)
                                val decodedBitmap = ImageUtil.instance?.compressImage(decodedBytes, 256f, 256f)
                                mvpView?.showUserImage(decodedBitmap)
                            } catch (e: IOException) {
                                Log.d("userimage", e.toString())
                            }
                        }
                    })?.let {
                        compositeDisposable?.add(it
                        )
                    }
        }
    val unreadNotificationsCount: Unit
        get() {
            compositeDisposable?.add(dataManager.unreadNotificationsCount
                    .observeOn(AndroidSchedulers.mainThread())
                    .subscribeOn(Schedulers.computation())
                    .subscribeWith(object : DisposableObserver<Int?>() {
                        override fun onComplete() {}
                        override fun onError(e: Throwable) {}
                        override fun onNext(integer: Int) {
                            mvpView?.showNotificationCount(integer)
                        }
                    }))
        }

}